### PR TITLE
Update to latest .NET 7 Roslyn compiler

### PR DESCRIPTION
**Pull request by jaredpar**
Several recent changes in the Roslyn compiler required updates to the runtime code base. These changes are mostly suppressing warnings / adding `scoped`. It's necessary to merge them into `main` and then backport to `release/7.0` in order to get source build clean.

### CODE REVIEW

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/DynamicMethod.CoreCLR.cs [hunk 1]
```diff
@@ -172,7 +172,9 @@ Signature LazyCreateSignature()
                     Span<ParameterCopyBackAction> shouldCopyBackParameters = new(ref argStorage._copyBack0, argCount);
 
                     StackAllocatedByRefs byrefStorage = default;
+                    #pragma warning disable CS8500
                     IntPtr* pByRefStorage = (IntPtr*)&byrefStorage;
+                    #pragma warning restore CS8500
```
Comment (stephentoub) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/DynamicMethod.CoreCLR.cs:177]: Nit: the other pragmas are left justified

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/DynamicMethod.CoreCLR.cs [hunk 2]
```diff
@@ -172,7 +172,9 @@ Signature LazyCreateSignature()
                     Span<ParameterCopyBackAction> shouldCopyBackParameters = new(ref argStorage._copyBack0, argCount);
 
                     StackAllocatedByRefs byrefStorage = default;
+                    #pragma warning disable CS8500
```
Comment (stephentoub) [inline src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/DynamicMethod.CoreCLR.cs:175]: What is CS8500?  I just want to understand what the compiler is concerned about.  I'm surprised taking the address of anything would produce a warning, regardless of where it came from.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.String.cs [hunk 3]
```diff
@@ -112,12 +112,12 @@ private void WriteStringEscapeProperty(scoped ReadOnlySpan<char> propertyName, i
             Debug.Assert(int.MaxValue / JsonConstants.MaxExpansionFactorWhileEscaping >= propertyName.Length);
 
             char[]? propertyArray = null;
+            scoped Span<char> escapedPropertyName;
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.String.cs:115]: For my edification, why is it necessary to move this out of the if block?

File: src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/DynamicMethod.CoreCLR.cs (hunk 2 above)
Comment (jaredpar) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/DynamicMethod.CoreCLR.cs:175]: `StackAllocatedByRefs` is a managed type by virtue of having `ref` fields. The warning here is fallout from the LDM decision of downgrading memory safety issues from errors to warnings inside an unsafe block. Taking the address of a managed type has traditionally been an error but now it's a warning when done inside unsafe blocks.

Comment (stephentoub) [reply src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/DynamicMethod.CoreCLR.cs:175]: Aha. Ok.

File: src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.String.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.String.cs:115]: > Once we fixed the bug to have scoped tie to the block

Got it.

**Comment by jaredpar** [comment]
@stephentoub are we good to merge here? The OSX failure seems to be just an Azure Agent disconnecting during execution.

**Review by stephentoub — APPROVED** [review]

File: src/coreclr/tools/aot/ILCompiler.Compiler.Tests/ILCompiler.Compiler.Tests.Assets/ILCompiler.Compiler.Tests.Assets.csproj [hunk 4]
```diff
@@ -8,6 +8,7 @@
     <TargetFramework>netstandard2.0</TargetFramework>
     <!-- Don't add references to the netstandard platform since this is a core assembly -->
     <DisableImplicitFrameworkReferences>true</DisableImplicitFrameworkReferences>
+    <Features>noRefSafetyRulesAttribute=true</Features>
```
Comment (stephentoub) [inline src/coreclr/tools/aot/ILCompiler.Compiler.Tests/ILCompiler.Compiler.Tests.Assets/ILCompiler.Compiler.Tests.Assets.csproj:11]: What does this do?

Comment (jaredpar) [reply src/coreclr/tools/aot/ILCompiler.Compiler.Tests/ILCompiler.Compiler.Tests.Assets/ILCompiler.Compiler.Tests.Assets.csproj:11]: This disables the compiler emitting the `[assembly: RefSafetyRules]` attribute + supporting type. The core attribute types are incorrectly defined in this project and by default results in a compilation error. Given this change is being back ported to `release/7.0` I opted for disabling the attribute emit vs. churning the code further. @cston and I were going to follow up with a PR in a bit that flips this off, fixes the attributes and let that play out separately.

**Pull request merged**